File: adventofcode2020/d3.cpp
// with O2 optimization flag, execution time is 0.005s
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
using namespace std;
vector<string> inputs;
int f(int r, int d) {
    int count=0,xinput=0,yinput=0;
    while (yinput<inputs.size()) {
        char pos=inputs[yinput][xinput%inputs[0].length()];
        if (pos=='#') count++;
        xinput+=r;
        yinput+=d;
    }
    return count;
}
int main() {
    ifstream in("i3");
    string raw_input;
    while (in >> raw_input) inputs.push_back(raw_input);
    // part 1 answer 257, part 2 answer 1744787392
    cout << f(3,1) << '\n' << f(1,1)*f(3,1)*f(5,1)*f(7,1)*f(1,2) << '\n';
}

File: adventofcode2020/d3.py
file=open("i3", 'r')
inputs=[x[:-1] for x in file.readlines()]
file.close()

#the first index is the y value, the second index is the x value.
# evaluate r3d1 which is right 3, down 1
def f(r, d):
    count=0
    xinput=0
    yinput=0
    while yinput<=len(inputs)-1:
        # record pos
        try:
            pos=inputs[yinput][xinput]
        except:
            xinput-=len(inputs[0])
            pos=inputs[yinput][xinput]
        if pos=='#': count+=1
        xinput+=r
        yinput+=d
    return count
# Part 1 answer 257
print(f(3,1))
# Part 2 answer 1744787392
print(f(1,1)*f(3,1)*f(5,1)*f(7,1)*f(1,2))